pennylane/ops/op_math/prod.py

```diff
@@ -418,3 +430,150 @@ def _swappable_ops(op1, op2, wire_map: dict = None) -> bool:
     if np.intersect1d(wires1, wires2).size != 0:
         return False
     return np.min(wires1) > np.min(wires2)
+
+
+class ProductFactorsGrouping:
+    """Utils class used for grouping identical product factors."""
+
+    _pauli_mult = {
+        "II": (1.0, "I"),
+        "IX": (1.0, "X"),
+        "IY": (1.0, "Y"),
+        "IZ": (1.0, "Z"),
+        "XI": (1.0, "X"),
+        "XX": (1.0, "I"),
+        "XY": (1.0j, "Z"),
+        "XZ": (-1.0j, "Y"),
+        "YI": (1.0, "Y"),
+        "YX": (-1.0j, "Z"),
+        "YY": (1.0, "I"),
+        "YZ": (1.0j, "X"),
+        "ZI": (1.0, "Z"),
+        "ZX": (1.0j, "Y"),
+        "ZY": (-1.0j, "X"),
+        "ZZ": (1.0, "I"),
+    }
+    _paulis = {"X": PauliX, "Y": PauliY, "Z": PauliZ}
+
+    def __init__(self):
+        self._pauli_factors = {}  #  {wire: (pauli_coeff, pauli_word)}
+        self._non_pauli_factors = {}  # {wires: [hash, pow_coeff, operator]}
+        self.queue = ()
+        self.global_phase = 1
+
+    def add(self, factor: Operator):
+        """Add operator to the factors queue.
+
+        Args:
+            factor (Operator): Operator to add to the queue.
+        """
+        if isinstance(factor, Prod):
+            for prod_factor in factor.factors:
+                self.add(prod_factor)
+        elif isinstance(factor, Sum):
+            self.queue += (factor.summands,)
+        elif not isinstance(factor, qml.Identity):
+            if isinstance(factor, SProd):
+                self.global_phase *= factor.scalar
+                factor = factor.base
+            wires = factor.wires
+            if isinstance(factor, (qml.Identity, qml.PauliX, qml.PauliY, qml.PauliZ)):
+                self._add_pauli_factor(factor=factor, wires=wires)
+                self._queue_non_pauli_factors(wires=wires)
```

Comment:
Add Pauli factor to the `self._pauli_fators` dictionary.

Remove all non Pauli factors that act on the same wires from the `self._non_pauli_factors` dictionary.